storage-proofs/src/crypto/feistel.rs

```diff
@@ -99,27 +102,40 @@ fn decode(index: u32, keys: &[u32], precomputed: FeistelPrecomputed) -> u32 {
     (left << half_bits) | right
 }
 
+const HALF_FEISTEL_BYTES: usize = mem::size_of::<Index>();
+const FEISTEL_BYTES: usize = 2 * HALF_FEISTEL_BYTES;
+
 // Round function of the Feistel network: `F(Ri, Ki)`. Joins the `right`
 // piece and the `key`, hashes it and returns the lower `u32` part of
 // the hash filtered trough the `right_mask`.
-fn feistel(right: u32, key: u32, right_mask: u32) -> u32 {
-    let mut data: [u8; 8] = [0; 8];
-    data[0] = (right >> 24) as u8;
-    data[1] = (right >> 16) as u8;
-    data[2] = (right >> 8) as u8;
-    data[3] = right as u8;
+#[allow(clippy::needless_range_loop)]
+fn feistel(right: Index, key: Index, right_mask: Index) -> Index {
+    let mut data: [u8; FEISTEL_BYTES] = [0; FEISTEL_BYTES];
+
+    {
+        let mut shift = (HALF_FEISTEL_BYTES - 1) * 8;
+
+        for item in data.iter_mut().take(HALF_FEISTEL_BYTES) {
+            *item = (right >> shift) as u8;
+            if shift > 0 {
+                shift -= 8;
+            }
+        }
+    }
 
-    data[4] = (key >> 24) as u8;
-    data[5] = (key >> 16) as u8;
-    data[6] = (key >> 8) as u8;
-    data[7] = key as u8;
+    {
+        let mut shift = (HALF_FEISTEL_BYTES - 1) * 8;
+        for i in 0..HALF_FEISTEL_BYTES {
+            data[i] = (key >> shift) as u8;
+            if shift > 0 {
+                shift -= 8;
+            }
+        }
+    }
 
-    let hash = Blake2s::digest(&data);
+    let hash = FeistelHash::digest(&data);
 
-    let r = u32::from(hash[0]) << 24
-        | u32::from(hash[1]) << 16
-        | u32::from(hash[2]) << 8
-        | u32::from(hash[3]);
+    let r = (0..HALF_FEISTEL_BYTES).fold(0, |acc, i| acc | Index::from(hash[i * 8]));
```

Comment:
this seems to be doing sth different than before, or at least looks much slower

Reply:
Yes, looks like I mangled something. Thanks — will push a fix.

Reply:
@dignifiedquire Done.